cdap-watchdog/src/main/java/io/cdap/cdap/logging/ErrorLogsClassifier.java

```diff
@@ -140,8 +161,17 @@ private List<ErrorClassificationResponse> getErrorClassificationResponse(String
         throwableProxy = throwableProxy.getCause();
       }
     }
+
     List<ErrorClassificationResponse> responses = new ArrayList<>(responseMap.values());
     responses.addAll(responseSet);
+
+    if (responses.isEmpty()) {
+      // use rule-based classification
+      ErrorClassificationResponse response = findMatchingRule(logIter2);
```

Comment:
I would consider doing it in one scan. Just extract this two pieces of logic (inside `while`) into separate methods / builder classes and call both for each LogEvent. You will do both work in all cases, but it's cheap work. And scanning is expensive.

Reply:
done.